src/Optimizer/Transformer/AutoExtensions.php

```diff
@@ -0,0 +1,609 @@
+                $attributeNameWithoutBindPrefix = substr($attribute->name, strlen(self::BIND_SHORT_FORM_PREFIX));
+
+                // Change the name from aria-labelledby to [ARIA_LABELLEDBY].
+                $bindableAttributeName = '['
+                    . preg_replace('/\-/', '_', strtoupper($attributeNameWithoutBindPrefix))
+                    . ']';
+
+                // Rename attribute from bindx to data-amp-bind-x.
+                if ($globalAttributes->has($bindableAttributeName)) {
+                    $newAttributeName = Amp::BIND_DATA_ATTR_PREFIX . $attributeNameWithoutBindPrefix;
+                    $newAttributes[$newAttributeName] = $attribute->value;
+
+                    $node->removeAttribute($attribute->name);
+
+                    $usesAmpBind = true;
+                }
+            }
+        }
+
+        // Add the renamed bindable attributes in the node.
+        if ($newAttributes) {
+            $node->setAttributes($newAttributes);
+        }
+
+        if ($usesAmpBind) {
+            $extensionScripts = $this->maybeAddExtension($document, $extensionScripts, Extension::BIND);
+        }
+
+        return $extensionScripts;
+    }
+
+    /**
+     * Get all attributes for a tag.
+     *
+     * @TODO: Provide this functionality as part of the validator spec implementation.
+     *
+     * @param string $tagName Name of the node tag.
+     * @return array Attribute list for the tag name.
+     */
+    private function getTagAttributeList($tagName)
+    {
+        static $nodeAttributeList = [];
+
+        if (! isset($nodeAttributeList[$tagName])) {
+            $nodeAttributeList[$tagName] = [];
+
+            $tagSpecs = $this->spec->tags()->byTagName($tagName);
+
+            foreach ($tagSpecs as $tagSpec) {
+                $attrLists = $tagSpec->get(SpecRule::ATTR_LISTS);
+                if (is_array($attrLists)) {
+                    foreach ($attrLists as $attrList) {
+                        $list = $this->spec->attributeLists()->get($attrList);
+                        $nodeAttributeList[$tagName] = array_merge($nodeAttributeList[$tagName], $list::ATTRIBUTES);
+                    }
+                }
+            }
+        }
+
+        return $nodeAttributeList[$tagName];
+    }
+
+    /**
+     * Remove unneeded extension scripts.
+     *
+     * @param Element[] $extensionScripts Array of extension scripts to check.
+     * @return Element[] Adapted array of extension scripts.
+     */
+    private function removeUnneededExtensions($extensionScripts)
+    {
+        if (! $this->configuration->get(AutoExtensionsConfiguration::REMOVE_UNNEEDED_EXTENSIONS)) {
+            return $extensionScripts;
+        }
+
+        return array_filter($extensionScripts, function ($extension) {
+            return array_key_exists($extension, $this->addedExtensions)
+                || in_array($extension, $this->protectedExtensions);
+        }, ARRAY_FILTER_USE_KEY);
```

Comment:
This logic needs to be adapted to only protect extensions if their dependencies are present.